force-app/main/default/classes/DMLOperations.cls

```diff
@@ -179,29 +280,61 @@ public with sharing class DMLOperations {
      * Example Contact.LastName = 'Jane' Link to Account.Name = 'Jane'
 	 * @param contactsToUpsert List of Contacts to be upserted and linked to accounts.
 	 */
-	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) {
+	public static void upsertAccountsWithContacts(List<Contact> contactsToUpsert) 
+    {
+        // Check if the list of contacts is not null or empty
+    if (contactsToUpsert == null || contactsToUpsert.isEmpty()) 
+    {
+        return; // Exit if input is invalid
+    }
 		// Loop through each Contact
-
-		// Extract the Account Name from Contact's LastName
-
-		// Get the account based on the name or create one if it doesn't exist
-
-		// Link the Contact to its Account using the Account Id
-
-		// Upsert the Contacts
+        for (Contact contact : contactsToUpsert) 
+        {
+            // Extract the Account Name from Contact's LastName
+            String accountName = contact.LastName;
+    
+            // Get the account based on the name or create one if it doesn't exist using the previous upsertAccount method
+            Account associatedAccount = upsertAccount(accountName);
```

Comment:
We will let this slide because the instructions say to do this, but calling methods that do DML within a loop also is bad practice.